File: react-native-poc/ios/RakeenPOC/RakeenPrinterModule.swift
import Foundation
import React

/// The real React Native NativeModule for printing — this is what
/// `NativeModules.RakeenPrinterModule` in
/// react-native-poc/src/platform/printer.ts resolves to on iOS. Exposed to
/// RN via RakeenPrinterModule.m (RCT_EXTERN_MODULE — the standard
/// classic-native-module bridging file; React Native's New Architecture,
/// enabled by default in this scaffold, still runs classic modules like
/// this one through its TurboModule interop layer without requiring
/// Codegen for a first POC).
///
/// UNVERIFIED beyond compilation — never run against a real printer or
/// device. See docs/react-native-poc/phase4-ios.md for exactly what
/// "Verified" means for this file vs. what still needs real hardware.
@objc(RakeenPrinterModule)
class RakeenPrinterModule: NSObject {

    private let transport = NetworkPrinterTransport()
    private let bluetoothTransport = BluetoothPrinterTransport()

    @objc
    static func requiresMainQueueSetup() -> Bool { return false }

    /// Matches the target.host/target.port/escPosBase64/timeoutMs shape
    /// PrintJob in src/platform/printer.ts sends — port is required, never
    /// defaulted to 9100 here. Also accepts the flat {host,port,...} shape
    /// for backward compatibility with the original POC screen. Feature
    /// Parity Pass -- Bluetooth/USB: `target.transport` now also routes to
    /// BluetoothPrinterTransport for `bluetoothId`; `usb` genuinely has no
    /// iOS implementation (ExternalAccessory/MFi-only restriction) and
    /// reports that honestly rather than pretending to succeed.
    @objc
    func print(_ job: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        let targetDict = (job["target"] as? NSDictionary) ?? job
        guard let base64 = job["escPosBase64"] as? String, let bytes = Data(base64Encoded: base64) else {
            resolve(["ok": false, "error": "RENDER_FAILED"])
            return
        }
        let timeoutMs = (job["timeoutMs"] as? NSNumber)?.intValue ?? 8000
        let transportKind = targetDict["transport"] as? String ?? "network"

        switch transportKind {
        case "bluetooth":
            guard let bluetoothId = targetDict["bluetoothId"] as? String else {
                resolve(["ok": false, "error": "INVALID_TARGET"])
                return
            }
            bluetoothTransport.send(bytes: bytes, peripheralId: bluetoothId, timeoutMs: timeoutMs) { ok, errorDetail in
                if ok {
                    resolve(["ok": true])
                } else if errorDetail == BluetoothPrinterTransport.TransportError.permissionDenied.rawValue {
                    resolve(["ok": false, "error": "PERMISSION_DENIED", "errorDetail": errorDetail ?? ""])
                } else {
                    resolve(["ok": false, "error": "PRINTER_CONNECTION_FAILED", "errorDetail": errorDetail ?? "unknown_error"])
                }
            }
        case "usb":
            resolve(["ok": false, "error": "TRANSPORT_NOT_SUPPORTED", "errorDetail": "usb_not_supported_on_ios"])
        default:
            guard let host = targetDict["host"] as? String, let portNumber = targetDict["port"] as? NSNumber else {
                resolve(["ok": false, "error": "INVALID_TARGET"])
                return
            }
            let port = UInt16(truncating: portNumber)
            transport.send(bytes: bytes, host: host, port: port) { ok, errorDetail in
                if ok {
                    resolve(["ok": true])
                } else {
                    resolve(["ok": false, "error": "PRINTER_CONNECTION_FAILED", "errorDetail": errorDetail ?? "unknown_error"])
                }
            }
        }
    }

    /// target: { transport, host?, port?, bluetoothId? }
    @objc
    func testConnection(_ target: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        let transportKind = target["transport"] as? String ?? "network"
        switch transportKind {
        case "bluetooth":
            guard let bluetoothId = target["bluetoothId"] as? String else {
                resolve(["reachable": false, "error": "INVALID_TARGET"])
                return
            }
            bluetoothTransport.testConnection(peripheralId: bluetoothId, timeoutMs: 8000) { reachable, latencyMs, errorDetail in
                var result: [String: Any] = ["reachable": reachable]
                if let latencyMs = latencyMs { result["latencyMs"] = latencyMs }
                if let errorDetail = errorDetail {
                    result["error"] = "PRINTER_CONNECTION_FAILED"
                    result["errorDetail"] = errorDetail
                }
                resolve(result)
            }
        case "usb":
            resolve(["reachable": false, "error": "TRANSPORT_NOT_SUPPORTED"])
        default:
            guard let host = target["host"] as? String, let portNumber = target["port"] as? NSNumber else {
                resolve(["reachable": false, "error": "INVALID_TARGET"])
                return
            }
            let port = UInt16(truncating: portNumber)
            transport.testConnection(host: host, port: port) { reachable, latencyMs, errorDetail in
                var result: [String: Any] = ["reachable": reachable]
                if let latencyMs = latencyMs { result["latencyMs"] = latencyMs }
                if let errorDetail = errorDetail {
                    result["error"] = "PRINTER_CONNECTION_FAILED"
                    result["errorDetail"] = errorDetail
                }
                resolve(result)
            }
        }
    }

    /// transport: 'network' | 'bluetooth' | 'usb', timeoutMs: Int.
    /// 'network' and 'usb' (no iOS USB host access) always resolve to an
    /// empty array -- never a fabricated device list.
    @objc
    func scanDevices(_ transportKind: String, timeoutMs: NSNumber, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        guard transportKind == "bluetooth" else {
            resolve(["devices": []])
            return
        }
        bluetoothTransport.scan(timeoutMs: timeoutMs.intValue) { devices, error in
            let mapped = devices.map { device -> [String: Any] in
                ["id": device.id, "rssi": device.rssi, "name": device.name ?? NSNull()]
            }
            var result: [String: Any] = ["devices": mapped]
            if let error = error {
                result["error"] = error == .permissionDenied ? "PERMISSION_DENIED" : "PRINTER_UNAVAILABLE"
            }
            resolve(result)
        }
    }

    /// Honest per docs/ios-native-bridge-interfaces.md's known gap: no
    /// bidirectional ESC/POS status read-back exists anywhere in this
    /// project yet, on either the Capacitor/Swift side or here. Always
    /// reports "unknown" rather than fabricating an "idle"/"ready" status
    /// nothing actually confirmed.
    @objc
    func getStatus(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        resolve(["status": "unknown"])
    }

    /// 'usb' is genuinely absent -- iOS has no USB host API reachable by a
    /// normal (non-MFi-certified) app for a generic ESC/POS printer, a
    /// real Apple platform restriction, not an unimplemented TODO.
    @objc
    func getCapabilities(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        resolve([
            "supportedTransports": ["network", "bluetooth"],
            "supportsCut": true,
            "supportsCashDrawerKick": true,
            "paperWidthPx": 576,
        ])
    }
}

File: react-native-poc/android/app/src/main/java/com/rakeenpoc/RakeenPrinterModule.kt
package com.rakeenpoc

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import android.util.Base64

/**
 * The real React Native NativeModule for printing on Android -- this is
 * what NativeModules.RakeenPrinterModule in
 * react-native-poc/src/platform/printer.ts resolves to on Android.
 * Registered via RakeenPackage.kt. Same JS-facing method names/shapes as
 * ios/RakeenPOC/RakeenPrinterModule.swift, deliberately -- that symmetry
 * IS the point of this whole POC (see docs/react-native-poc/phase2-architecture.md).
 *
 * UNVERIFIED beyond compilation -- never run against a real printer or a
 * real device. See docs/react-native-poc/phase5-android.md.
 */
class RakeenPrinterModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private val transport = NetworkPrinterTransport()
    private val bluetoothTransport = BluetoothClassicPrinterTransport(reactContext)
    private val usbTransport = UsbPrinterTransport(reactContext)

    override fun getName() = "RakeenPrinterModule"

    /** job: { target: { transport, host?, port?, bluetoothId?, usbAccessoryId? },
     *  escPosBase64: String, timeoutMs: Int }. Feature Parity Pass --
     *  Bluetooth/USB: `target.transport` routes to the real classic-
     *  Bluetooth (RFCOMM) or USB host transport; network keeps its
     *  original host/port-required behavior unchanged, matching iOS. */
    @ReactMethod
    fun print(job: ReadableMap, promise: Promise) {
        val target = if (job.hasKey("target")) job.getMap("target") else job
        val base64 = if (job.hasKey("escPosBase64")) job.getString("escPosBase64") else null
        val timeoutMs = if (job.hasKey("timeoutMs")) job.getInt("timeoutMs") else 8000

        val bytes = base64?.let {
            try {
                Base64.decode(it, Base64.DEFAULT)
            } catch (e: IllegalArgumentException) {
                null
            }
        }
        if (bytes == null) {
            val result = Arguments.createMap()
            result.putBoolean("ok", false)
            result.putString("error", "RENDER_FAILED")
            promise.resolve(result)
            return
        }

        val transportKind = target?.let { if (it.hasKey("transport")) it.getString("transport") else null } ?: "network"
        when (transportKind) {
            "bluetooth" -> {
                val bluetoothId = target?.let { if (it.hasKey("bluetoothId")) it.getString("bluetoothId") else null }
                if (bluetoothId == null) {
                    resolveInvalidTarget(promise)
                    return
                }
                bluetoothTransport.send(bluetoothId, bytes) { ok, errorDetail ->
                    resolvePrintResult(promise, ok, errorDetail)
                }
            }
            "usb" -> {
                val usbAccessoryId = target?.let { if (it.hasKey("usbAccessoryId")) it.getString("usbAccessoryId") else null }
                if (usbAccessoryId == null) {
                    resolveInvalidTarget(promise)
                    return
                }
                usbTransport.send(usbAccessoryId, bytes, timeoutMs) { ok, errorDetail ->
                    resolvePrintResult(promise, ok, errorDetail)
                }
            }
            else -> {
                val host = target?.let { if (it.hasKey("host")) it.getString("host") else null }
                val port = target?.let { if (it.hasKey("port")) it.getInt("port") else -1 } ?: -1
                if (host == null || port <= 0) {
                    resolveInvalidTarget(promise)
                    return
                }
                transport.send(host, port, bytes) { ok, errorDetail ->
                    resolvePrintResult(promise, ok, errorDetail)
                }
            }
        }
    }

    private fun resolveInvalidTarget(promise: Promise) {
        val result = Arguments.createMap()
        result.putBoolean("ok", false)
        result.putString("error", "INVALID_TARGET")
        promise.resolve(result)
    }

    private fun resolvePrintResult(promise: Promise, ok: Boolean, errorDetail: String?) {
        val result = Arguments.createMap()
        result.putBoolean("ok", ok)
        if (errorDetail != null) {
            result.putString("error", if (errorDetail == "permission_denied") "PERMISSION_DENIED" else "PRINTER_CONNECTION_FAILED")
            result.putString("errorDetail", errorDetail)
        }
        promise.resolve(result)
    }

    /** target: { transport, host?, port?, bluetoothId?, usbAccessoryId? } */
    @ReactMethod
    fun testConnection(target: ReadableMap, promise: Promise) {
        val transportKind = if (target.hasKey("transport")) target.getString("transport") else "network"
        when (transportKind) {
            "bluetooth" -> {
                val bluetoothId = if (target.hasKey("bluetoothId")) target.getString("bluetoothId") else null
                if (bluetoothId == null) { resolveInvalidConnectionTest(promise); return }
                bluetoothTransport.testConnection(bluetoothId) { reachable, latencyMs, errorDetail ->
                    resolveConnectionTest(promise, reachable, latencyMs, errorDetail)
                }
            }
            "usb" -> {
                val usbAccessoryId = if (target.hasKey("usbAccessoryId")) target.getString("usbAccessoryId") else null
                if (usbAccessoryId == null) { resolveInvalidConnectionTest(promise); return }
                usbTransport.testConnection(usbAccessoryId) { reachable, latencyMs, errorDetail ->
                    resolveConnectionTest(promise, reachable, latencyMs, errorDetail)
                }
            }
            else -> {
                val host = if (target.hasKey("host")) target.getString("host") else null
                val port = if (target.hasKey("port")) target.getInt("port") else -1
                if (host == null || port <= 0) { resolveInvalidConnectionTest(promise); return }
                transport.testConnection(host, port) { reachable, latencyMs, errorDetail ->
                    resolveConnectionTest(promise, reachable, latencyMs, errorDetail)
                }
            }
        }
    }

    private fun resolveInvalidConnectionTest(promise: Promise) {
        val result = Arguments.createMap()
        result.putBoolean("reachable", false)
        result.putString("error", "INVALID_TARGET")
        promise.resolve(result)
    }

    private fun resolveConnectionTest(promise: Promise, reachable: Boolean, latencyMs: Double?, errorDetail: String?) {
        val result = Arguments.createMap()
        result.putBoolean("reachable", reachable)
        if (latencyMs != null) result.putDouble("latencyMs", latencyMs)
        if (errorDetail != null) {
            result.putString("error", if (errorDetail == "permission_denied") "PERMISSION_DENIED" else "PRINTER_CONNECTION_FAILED")
            result.putString("errorDetail", errorDetail)
        }
        promise.resolve(result)
    }

    /** transport: 'network' | 'bluetooth' | 'usb', timeoutMs: Int.
     *  'network' has no discovery concept -- always resolves an empty list,
     *  never an error (nothing went wrong, there's just nothing to scan). */
    @ReactMethod
    fun scanDevices(transportKind: String, timeoutMs: Int, promise: Promise) {
        when (transportKind) {
            "bluetooth" -> bluetoothTransport.scan(timeoutMs) { devices, error ->
                val result = Arguments.createMap()
                val list = Arguments.createArray()
                devices.forEach { d ->
                    val entry = Arguments.createMap()
                    entry.putString("id", d.id)
                    if (d.name != null) entry.putString("name", d.name) else entry.putNull("name")
                    list.pushMap(entry)
                }
                result.putArray("devices", list)
                if (error != null) result.putString("error", if (error == "permission_denied") "PERMISSION_DENIED" else "PRINTER_UNAVAILABLE")
                promise.resolve(result)
            }
            "usb" -> usbTransport.scan { devices ->
                val result = Arguments.createMap()
                val list = Arguments.createArray()
                devices.forEach { d ->
                    val entry = Arguments.createMap()
                    entry.putString("id", d.id)
                    if (d.name != null) entry.putString("name", d.name) else entry.putNull("name")
                    list.pushMap(entry)
                }
                result.putArray("devices", list)
                promise.resolve(result)
            }
            else -> {
                val result = Arguments.createMap()
                result.putArray("devices", Arguments.createArray())
                promise.resolve(result)
            }
        }
    }

    /** Honest per docs/ios-native-bridge-interfaces.md's known gap: no
     *  bidirectional ESC/POS status read-back exists on either platform
     *  yet -- always reports "unknown", never a fabricated "idle"/"ready". */
    @ReactMethod
    fun getStatus(promise: Promise) {
        val result = Arguments.createMap()
        result.putString("status", "unknown")
        promise.resolve(result)
    }

    @ReactMethod
    fun getCapabilities(promise: Promise) {
        val result = Arguments.createMap()
        val transports = Arguments.createArray()
        transports.pushString("network")
        transports.pushString("bluetooth")
        transports.pushString("usb")
        result.putArray("supportedTransports", transports)
        result.putBoolean("supportsCut", true)
        result.putBoolean("supportsCashDrawerKick", true)
        result.putInt("paperWidthPx", 576)
        promise.resolve(result)
    }
}
